tendermint/src/lite/verifier.rs

```diff
@@ -260,30 +260,25 @@ where
 
     // inner recursive function which assumes
     // trusting_period check is already done.
-    _verify_and_update_bisection(untrusted_height, trust_threshold, req, store)
+    let new_trusted =
+        verify_and_update_bisection_inner(trusted_state, untrusted_height, trust_threshold, req)?;
+    store.add(new_trusted)
 }
 
 // inner recursive function for verify_and_update_bisection.
 // see that function's docs.
-fn _verify_and_update_bisection<H, C, L, R, S>(
+fn verify_and_update_bisection_inner<H, C, L, R>(
+    trusted_state: &TrustedState<C, H>,
     untrusted_height: Height,
     trust_threshold: &L,
     req: &R,
-    store: &mut S,
-) -> Result<(), Error>
+) -> Result<TrustedState<C, H>, Error>
```

Comment:
imho some explanation in the comments on the new returned value(s) would be useful. Also, I thought this should return a list of trusted states that at the end of the recursion will be added to the store by the caller.

Reply:
Thanks @ancazamfir. Yeah, I wasn't sure. But yes a list would be exactly the same behaviour as before! Will do. Regarding the comment: this is just a non-public helper method. The public one (`verify_bisection`) has a brief explanation but yes, these could be better too.

Reply:
Haha, today I learned there is a whole crate dealing with memoization: https://github.com/jaemk/cached
It provides a macro for things like this: https://medium.com/swlh/on-memoization-291fd1dd924
Interesting, but I likely just use a HashSet / HashMap or even a simple Vec (have to think about the latter, as we might request headers twice if there is some overlap?).

Reply:
> Haha, today I learned there is a whole crate dealing with memoization: https://github.com/jaemk/cached
It provides a macro for things like this: https://medium.com/swlh/on-memoization-291fd1dd924

Interesting!

> ....or even a simple Vec (have to think about the latter, as we might request headers twice if there is some overlap?).

Vec is nice as it will have the headers in the height order. I don't think we ask for headers twice...or at least we shouldn't.

Reply:
Made it a HashMap (used more like a HashSet) in 48a7a37. Changing this to a Vec and checking that we don't have double entries in there sounds like a good first test then.

Reply:
Changed to Vec in ed8c06e. Will add tests in a bit.